### PR TITLE
Fix vb partial determinism

### LINKED ISSUE #7111: VB interfaces not emitted it a deterministic order across partial types
Consider the following Code

``` 
' file1.vb
Partial Class C 
  Implements I1
End Class

' file2.vb
Partial Class C
  Implements I2
End class
```

These interfaces should be emitted in source order (irrespective of determinism).  At this time they are not.

### PR DESCRIPTION
**Pull request by gafter**
Fixes #7111 and #7112 

Also tests C# to ensure the same issue does not arise there.

@dotnet/roslyn-compiler Please review

### CODE REVIEW

File: src/Compilers/VisualBasic/Portable/Symbols/Source/SourceMemberContainerTypeSymbol.vb [hunk 1]
```diff
@@ -1378,9 +1378,25 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
             Return _lazyLexicalSortKey
         End Function
 
+        Public Function LocationKey(ByVal l As Location) As Long
+            If Not l.IsInSource Then
+                Dim vbLocation = TryCast(l, VBLocation)
+                Return 1L << 62 + If(vbLocation Is Nothing, 0, vbLocation.PossiblyEmbeddedOrMySourceSpan.Start)
+            End If
+
+            Dim treeIndex = If(l.IsInSource,
```
Comment (gafter) [inline src/Compilers/VisualBasic/Portable/Symbols/Source/SourceMemberContainerTypeSymbol.vb]: This test is redundant with the previous `If` and will be removed.

**Comment by jaredpar** [comment]
:+1: 

Looks like you have a real compilation error.

**Comment by jaredpar** [comment]
@TyOverby please review

File: src/Compilers/VisualBasic/Portable/Symbols/Source/SourceMemberContainerTypeSymbol.vb [hunk 2]
```diff
@@ -1378,9 +1378,23 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
             Return _lazyLexicalSortKey
         End Function
 
+        Public Function LocationKey(ByVal l As Location) As Long
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Symbols/Source/SourceMemberContainerTypeSymbol.vb]: Consider adding a comment about how do we fill certain bits in the key so that numbers 62 and 40 didn't look like something random.

File: src/Compilers/VisualBasic/Portable/Symbols/Source/SourceMemberContainerTypeSymbol.vb [hunk 3]
```diff
@@ -1378,9 +1378,23 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
...
+            If Not l.IsInSource Then
+                Dim vbLocation = TryCast(l, VBLocation)
+                Return 1L << 62 + If(vbLocation Is Nothing, 0, vbLocation.PossiblyEmbeddedOrMySourceSpan.Start)
+            End If
+
+            Dim treeIndex = Me.DeclaringCompilation.GetSyntaxTreeOrdinal(l.SourceTree)
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Symbols/Source/SourceMemberContainerTypeSymbol.vb]: Consider spelling out the type in order to improve readability of the code. I believe the general rule is - the type must be named unless it is named in the initializer.

File: src/Compilers/VisualBasic/Portable/Symbols/Source/SourceMemberContainerTypeSymbol.vb [hunk 4]
```diff
@@ -1378,9 +1378,23 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
             Return _lazyLexicalSortKey
         End Function
 
+        Public Function LocationKey(ByVal l As Location) As Long
+            If Not l.IsInSource Then
+                Dim vbLocation = TryCast(l, VBLocation)
+                Return 1L << 62 + If(vbLocation Is Nothing, 0, vbLocation.PossiblyEmbeddedOrMySourceSpan.Start)
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Symbols/Source/SourceMemberContainerTypeSymbol.vb]: Should we be using different prefix for different embedded trees?

File: src/Compilers/VisualBasic/Portable/Compilation/VisualBasicCompilation.vb [hunk 5]
```diff
@@ -785,7 +785,7 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
             Get
                 If _lazyAllSyntaxTrees.IsDefault Then
                     Dim builder = ArrayBuilder(Of SyntaxTree).GetInstance()
-                    builder.AddRange(_rootNamespaces.Keys)
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Compilation/VisualBasicCompilation.vb]: I think we should rename _rootNamespaces. Its name is confusing because it doesn't contain information about embedded trees, but one might assume it does.

File: src/Compilers/VisualBasic/Portable/Symbols/Source/SourceMemberContainerTypeSymbol.vb [hunk 6]
```diff
@@ -1391,7 +1405,11 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
...
         Public ReadOnly Property SyntaxReferences As ImmutableArray(Of SyntaxReference)
             Get
-                Return _declaration.SyntaxReferences
+                Dim result = _declaration.SyntaxReferences
+                If result.Length >= 2 Then
+                    result = result.OrderBy(Function(ByVal r As SyntaxReference) LocationKey(r.GetLocation())).ToImmutableArray()
+                End If
+                Return result
             End Get
         End Property
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Symbols/Source/SourceMemberContainerTypeSymbol.vb]: It is not obvious how changes in this properties actually fixing the issues, could you provide some details?

Comment (gafter) [inline src/Compilers/VisualBasic/Portable/Symbols/Source/SourceMemberContainerTypeSymbol.vb]: We iterator over SyntaxReferences to process the parts. If SyntaxReferences are not in a deterministic order, then we will process the parts in an arbitrary order.

File: src/Compilers/VisualBasic/Portable/Compilation/VisualBasicCompilation.vb [hunk 7]
```diff
@@ -785,7 +785,7 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
...
+                    builder.AddRange(_syntaxTrees)
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Compilation/VisualBasicCompilation.vb]: What is the motivation behind this change? Based on the comment for ```SyntaxTrees``` property, it doesn't look like ```_syntaxTrees``` guarantee any specific order?

File: src/Compilers/VisualBasic/Portable/Symbols/Source/SourceMemberContainerTypeSymbol.vb (hunk 6 above)
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Symbols/Source/SourceMemberContainerTypeSymbol.vb]: Could you please point to the code where "We iterate over SyntaxReferences to process the parts."

File: src/Compilers/VisualBasic/Portable/Compilation/VisualBasicCompilation.vb (hunk 5 above)
Comment (gafter) [inline src/Compilers/VisualBasic/Portable/Compilation/VisualBasicCompilation.vb]: Great idea, but out of scope for this PR.

File: src/Compilers/VisualBasic/Portable/Compilation/VisualBasicCompilation.vb (hunk 7 above)
Comment (gafter) [inline src/Compilers/VisualBasic/Portable/Compilation/VisualBasicCompilation.vb]: The spec for `Compilation.SyntaxTrees` is that it is the sequence of trees the compilation was created with. I don't know why VB's property has a different specification that appears to allow nondeterminism. It is in fact deterministic, as it must be to have reliable semantics. `_syntaxTrees` is in the order provided to the compilation. `_rootNamespaces` on the other hand is a `Dictionary` and so its keys change order from compilation to compilation.

There are even assertions that the order is deterministic, such as this one in the constructor:

```vb
            Debug.Assert(syntaxTrees.All(Function(tree) syntaxTrees(syntaxTreeOrdinalMap(tree)) Is tree))
```

I'll remove the part of the spec that allows nondeterminism.

**Comment by gafter** [comment]
After some discussion with @AlekseyTs , I'm going to investigate why we're getting nondeterministic results from the declaration table. The C# side appears to be deterministic. If that were deterministic, we would not need compensating code above it. It is possible that `DeclaringReferences` is nondeterministic for namespace symbols; that too should be made deterministic.

Also @AlekseyTs mentioned that we already have comparators for sorting based on location; it would be better to reuse that mechanism if possible.

File: src/Compilers/VisualBasic/Portable/Symbols/Source/SourceMemberContainerTypeSymbol.vb (hunk 2 above)
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Symbols/Source/SourceMemberContainerTypeSymbol.vb]: We already have LexicalSortKey class and Compilation.CompareSourceLocations method that utilizes it. Consider taking advantage of them so that we get consistent sorting throughout the compiler. For example, GetBestName method in this class uses LexicalSortKey.

File: src/Compilers/VisualBasic/Portable/Symbols/Source/SourceMemberContainerTypeSymbol.vb (hunk 6 above)
Comment (gafter) [inline src/Compilers/VisualBasic/Portable/Symbols/Source/SourceMemberContainerTypeSymbol.vb]: Se https://github.com/dotnet/roslyn/blob/master/src/Compilers/VisualBasic/Portable/Symbols/Source/SourceNamedTypeSymbol.vb#L1308 for one example.

Comment (gafter) [inline src/Compilers/VisualBasic/Portable/Symbols/Source/SourceMemberContainerTypeSymbol.vb]: Though now that I look at it, it places the result in a `HashSet`!

**Comment by gafter** [comment]
@dotnet/roslyn-compiler OK, I've moved the main bug fix into the declaration table, which is now deterministic. Please re-review.

File: src/Compilers/Core/Portable/InternalUtilities/SetWithInsertionOrder.cs [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+    /// </summary>
+    internal sealed class SetWithInsertionOrder<T> : IEnumerable<T>
+    {
+        private Dictionary<T, uint> _map = new Dictionary<T, uint>();
+        private uint _nextElementValue = 0;
+        private T[] _elements = null;
+
+        public bool Add(T value)
+        {
+            if (_map.ContainsKey(value)) return false;
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/InternalUtilities/SetWithInsertionOrder.cs]: This could be optimized by the following 

```
if (!_map.Add(value))
  return false;

// rest of code
```

In the current form it uses `Contains` and `Add` which causes lookup to happen twice.

File: src/Compilers/Core/Portable/InternalUtilities/SetWithInsertionOrder.cs [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+            var thisValue = _nextElementValue++;
+            _map.Add(value, thisValue);
+            if (_elements == null)
+            {
+                _elements = new T[10];
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/InternalUtilities/SetWithInsertionOrder.cs]: Why 10 here?  Types like `List<T>` use 4 which is a bit smaller.

File: src/Compilers/Core/Portable/InternalUtilities/SetWithInsertionOrder.cs [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+namespace Roslyn.Utilities
+{
+    /// <summary>
+    /// A set that returns the inserted values in insertion order.
+    /// The mutation operations are not thread-safe.
+    /// </summary>
+    internal sealed class SetWithInsertionOrder<T> : IEnumerable<T>
+    {
+        private Dictionary<T, uint> _map = new Dictionary<T, uint>();
+        private uint _nextElementValue = 0;
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/InternalUtilities/SetWithInsertionOrder.cs]: Why `uint` here?  It seems every use, except for the increment, implicitly or explicitly converts to `int`

File: src/Compilers/Core/Portable/InternalUtilities/SetWithInsertionOrder.cs [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+                Array.Resize(ref _elements, _elements.Length * 2);
+            }
+
+            _elements[thisValue] = value;
+            return true;
+        }
+
+        public int Count => (int)_nextElementValue;
+
+        public bool Contains(T value) => _map.ContainsKey(value);
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/InternalUtilities/SetWithInsertionOrder.cs]: Shloud we use `ContainsKey` here to make it easier to transition `HashSet<T>` to `SetWithInsertionOrder<T>`?

File: src/Compilers/Core/Portable/InternalUtilities/SetWithInsertionOrder.cs [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+
+        public IEnumerator<T> GetEnumerator()
+        {
+            for (int i = 0; i < _nextElementValue; i++) yield return _elements[i];
+        }
+
+        IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+
+        public SetWithInsertionOrder<T> InInsertionOrder => this;
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/InternalUtilities/SetWithInsertionOrder.cs]: What is the purpose of this method?  Seems to just be an identity function.

File: src/Compilers/Core/Portable/InternalUtilities/SetWithInsertionOrder.cs (hunk 11 above)
Comment (gafter) [inline src/Compilers/Core/Portable/InternalUtilities/SetWithInsertionOrder.cs]: `HashSet<T>` doesn't have an operation called `ContainsKey`.

Comment (jaredpar) [inline src/Compilers/Core/Portable/InternalUtilities/SetWithInsertionOrder.cs]: Nevermind, `HashSet<T>` uses `Contains`.

File: src/Compilers/Core/Portable/InternalUtilities/SetWithInsertionOrder.cs [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+
+namespace Roslyn.Utilities
+{
+    /// <summary>
+    /// A set that returns the inserted values in insertion order.
+    /// The mutation operations are not thread-safe.
+    /// </summary>
+    internal sealed class SetWithInsertionOrder<T> : IEnumerable<T>
+    {
+        private Dictionary<T, uint> _map = new Dictionary<T, uint>();
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/InternalUtilities/SetWithInsertionOrder.cs]: Why `Dictionary` instead of `HashSet` here?

File: src/Compilers/Core/Portable/InternalUtilities/SetWithInsertionOrder.cs (hunk 10 above)
Comment (gafter) [inline src/Compilers/Core/Portable/InternalUtilities/SetWithInsertionOrder.cs]: Because it cannot be negative. This is analogous code to the immutable version.

File: src/Compilers/Core/Portable/InternalUtilities/SetWithInsertionOrder.cs (hunk 9 above)
Comment (gafter) [inline src/Compilers/Core/Portable/InternalUtilities/SetWithInsertionOrder.cs]: I stuck my fingers in the air and felt which way the wind was blowing. I felt the wind on all ten fingers.

Comment (jaredpar) [inline src/Compilers/Core/Portable/InternalUtilities/SetWithInsertionOrder.cs]: Love it :smile:

File: src/Compilers/Core/Portable/InternalUtilities/SetWithInsertionOrder.cs (hunk 8 above)
Comment (gafter) [inline src/Compilers/Core/Portable/InternalUtilities/SetWithInsertionOrder.cs]: `Dictionary.Add` returns `void` and throws an exception if the key is already present.

Comment (jaredpar) [inline src/Compilers/Core/Portable/InternalUtilities/SetWithInsertionOrder.cs]: Yeah I wrote that comment thinking `_map` was a `HashSet<T>`.  Seems like that is a more ideal type for this implementation.

File: src/Compilers/Core/Portable/InternalUtilities/SetWithInsertionOrder.cs (hunk 13 above)
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/InternalUtilities/SetWithInsertionOrder.cs]: Should we use SmallDictionary?

Comment (AlekseyTs) [inline src/Compilers/Core/Portable/InternalUtilities/SetWithInsertionOrder.cs]: Also, it looks like the value is never used. Consider making its type bool, in order to save some memory.

Comment (AlekseyTs) [inline src/Compilers/Core/Portable/InternalUtilities/SetWithInsertionOrder.cs]: LGTM

File: src/Compilers/Core/Portable/InternalUtilities/SetWithInsertionOrder.cs (hunk 8 above)
Comment (gafter) [inline src/Compilers/Core/Portable/InternalUtilities/SetWithInsertionOrder.cs]: Although I did not implement `Remove` yet (because I didn't need it), using a `HashSet` instead of a `Dictionary` would make it difficult to implement efficiently.

File: src/Compilers/Core/Portable/InternalUtilities/SetWithInsertionOrder.cs (hunk 12 above)
Comment (gafter) [inline src/Compilers/Core/Portable/InternalUtilities/SetWithInsertionOrder.cs]: It is to be consistent with the immutable version, where you have to call `IsInsertionOrder` to get a deterministic enumeration. We happen to yield insertion order both ways, but this way the user can specifically ask for it.

File: src/Compilers/Core/Portable/InternalUtilities/SetWithInsertionOrder.cs (hunk 13 above)
Comment (gafter) [inline src/Compilers/Core/Portable/InternalUtilities/SetWithInsertionOrder.cs]: My intent was that the dictionary's values would be used later to implement `Remove`, but that doesn't really work (once you remove something all the indices shift). So I'll switch to a `HashSet`.

**Comment by jaredpar** [comment]
:+1:

**Comment by TyOverby** [comment]
:+1:

**Comment by gafter** [comment]
closed in 487c201
